homeassistant/components/stream/core.py

```diff
@@ -42,17 +63,166 @@ class Segment:
     # For detecting discontinuities across stream restarts
     stream_id: int = attr.ib(default=0)
-    parts: list[Part] = attr.ib(factory=list)
+    # Parts are stored in a dict indexed by byterange for easy lookup
+    # As of Python 3.7, insertion order is preserved, and we insert
+    # in sequential order, so the Parts are ordered
+    parts_by_byterange: dict[int, Part] = attr.ib(factory=dict)
     start_time: datetime.datetime = attr.ib(factory=datetime.datetime.utcnow)
+    _stream_outputs: Iterable[StreamOutput] = attr.ib(factory=list)
+    # Store text of this segment's hls playlist for reuse
+    hls_playlist_template: str = attr.ib(default=None)
+    hls_playlist_parts: str = attr.ib(default=None)
+    # Number of playlist parts rendered so far
+    hls_num_parts_rendered: int = attr.ib(default=0)
+    # Set to true when all the parts are rendered
+    hls_playlist_complete: bool = attr.ib(default=False)
+
+    def __attrs_post_init__(self) -> None:
+        """Run after init."""
+        for output in self._stream_outputs:
+            output.put(self)
 
     @property
     def complete(self) -> bool:
         """Return whether the Segment is complete."""
         return self.duration > 0
 
-    def get_bytes_without_init(self) -> bytes:
+    @property
+    def data_size_with_init(self) -> int:
+        """Return the size of all part data + init in bytes."""
+        return len(self.init) + self.data_size
+
+    @property
+    def data_size(self) -> int:
+        """Return the size of all part data without init in bytes."""
+        # We can use the last part to quickly calculate the total data size.
+        if not self.parts_by_byterange:
+            return 0
+        last_http_range_start, last_part = next(
+            reversed(self.parts_by_byterange.items())
+        )
+        return last_http_range_start + len(last_part.data)
+
+    @callback
+    def async_add_part(
+        self,
+        part: Part,
+        duration: float,
+    ) -> None:
+        """Add a part to the Segment.
+
+        Duration is non zero only for the last part.
+        """
+        self.parts_by_byterange[self.data_size] = part
+        self.duration = duration
+        for output in self._stream_outputs:
+            output.part_put()
+
+    def get_data(self) -> bytes:
         """Return reconstructed data for all parts as bytes, without init."""
-        return b"".join([part.data for part in self.parts])
+        return b"".join([part.data for part in self.parts_by_byterange.values()])
+
+    def get_aggregating_bytes(
+        self, start_loc: int, end_loc: int | float
+    ) -> Generator[bytes, None, None]:
+        """Yield available remaining data until segment is complete or end_loc is reached.
+
+        Begin at start_loc. End at end_loc (exclusive).
+        Used to help serve a range request on a segment.
+        """
+        pos = start_loc
+        while (part := self.parts_by_byterange.get(pos)) or not self.complete:
```

Comment:
Does start_loc need to be exactly on a part boundary?

Reply:
Yes right now it does. This should be fine since the requests coming in should also be aligned with the part byteranges. We can add a comment to note this if there isn't one in there already.

Reply:
Added a comment to the docstring of `HlsSegmentView.handle()`